### ISSUE TITLE
remove_small_objects doesn't work with dtype=uint8 

### ISSUE THREAD

**Issue by OmriTreidel** [scikit-image/scikit-image#1620]
M  = array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)

It fails to remove the isolated 1 at location 2,2

**Comment by stefanv** [comment]
Works for me on 0.12dev.

**Comment by jni** [comment]
@OmriTreidel this is expected behaviour. From the docstring: "If the array type is int, it is assumed that it contains already-labeled objects." Therefore, the isolated 1 is considered part of the rest of the 1s. To remove it, you need to relabel it, or to pass the image as a bool image.

@stefanv that's surprising! Are you sure you passed in `uint8`?

**Comment by OmriTreidel** [comment]
Yes, the matrix I added is the printout, so its definitely uint8

**Comment by jni** [comment]
@OmriTreidel I know yours is, cos it doesn't work, as expected. =) I just meant I was surprised @stefanv got it to work with the exact same input.

**Comment by stefanv** [comment]
```
In [10]: M = array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
[0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
[0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
[0, 0, 0, 0, 0, 1, 1, 1, 1, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)

In [11]: morphology.remove_small_objects(M)
/home/stefan/src/scikit-image/skimage/morphology/misc.py:119: UserWarning: Only one label was provided to `remove_small_objects`. Did you mean to use a boolean array?
  warnings.warn("Only one label was provided to `remove_small_objects`. "
Out[11]: 
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)

In [12]: skimage.__version__
Out[12]: '0.12dev'
```

**Comment by jni** [comment]
@stefanv LOL That's because you're using the default size threshold of 64. What @OmriTreidel was asking is to remove only the `1` at (2, 2) using a size threshold of 2.

**Comment by stefanv** [comment]
Ah, makes sense!

This behavior is confusing, I think. Should we consider changing it?

**Comment by jni** [comment]
@stefanv it's there for a good reason, which is that people might want to use whatever fancy labelling algorithm they have and then remove the small resulting objects — which may be discontiguous, or not separated by 0s. It's also much more efficient not to re-run the labelling if it's already been done.

When there are just 0 and 1 as labels, we could detect it as a binary image — which we already do, see the warning you raised! It even tells you to convert your image to bool. Back in the day, automatically converting it was seen as too much magic. I myself like magic so would be ok with changing it.

Josh might argue that you don't want the behaviour to magically change with only one label, because you might be processing a series of images, and one of them will happen to have a single label and you want to treat it as such. There would be no way whatsoever for users to do a `remove_small_objects` on an image with a single label if we change the behaviour.

**Comment by stefanv** [comment]
I guess I wasn't clear: I was asking to remove any magic functionality, and
not to do selective interpretation of the input data. I hear the
effectivity argument, but really it is quite cheap to relabel an array.
Memory use may be a problem.

**Comment by jni** [comment]
Oh. Well then I'm 100% against this. =P "relabeling" means different things in different contexts. Often I want the behaviour of `ndi.label`, which treats the volume as boolean, and other times I want that of `measure.label`, which looks for contiguous segments of the same label. If we normalize to `ndi.label`, we lose the ability to remove small objects labeled with `measure.label`, and if we normalize to `measure.label`, we lose support for 4D+ images. Also, the current behaviour allows you to maintain your original labels, which can be important in many applications.

In short, I think it's a terrible idea.

**Comment by stefanv** [comment]
That'd rather emphatic, even for you Juan.

**Comment by JDWarner** [comment]
A discussion about the default size threshold (64) may be warranted.

Stepping back for a moment, how would one expect the algorithm to behave when faced with a small, unconnected region sufficiently far from a larger, same-labeled region such that it would be removed if they held different labels? Would two points with the same label, far apart, be considered a large object? What about three? Is it the convex hull that determines if disjoint sets are large enough to keep?

The docstring is silent on the topic.

If there are discrete use cases here, IMO they should be either hard switched with a kwarg or refactored into separate functions. Then you have deterministic behavior without sacrificing functionality.

**Comment by jni** [comment]
@JDWarner re the default, I picked it because it's an integer-sized in both 2D and 3D (8x8 or 4x4x4), and a reasonable size for many applications. But of course it's highly dependent on application. Perhaps we should not have a default at all.

Regarding your use case, you're being way too complicated! Convex hulls??? It's just 1 label = 1 region. If the number of pixels with that label is below a threshold, that label is removed.

The only exception to this rule is if a boolean volume is passed. Then, `ndi.label` is called on the volume before doing the exact same thing.
